### PR TITLE
feat(REST): dynamic rate limit offsets

### LINKED ISSUE #9762: Custom REST rate limit offsets for specific routes
### Which application or package is this feature request for?

rest

### Feature

To help prevent ratelimits and be more safe, discord.js gives you the option to set a ratelimit offset for all routes. But, some routes are heavily used by some bots, perhaps more than other routes, so providing a one size fits all may not be ideal all the time.

For example what if you have a reactions heavy bot, that adds reactions very often, and you want to be more cautious of the frequent reactions your bot is adding and want to avoid 429s, but you don't want it to make other actions slower than preferred, like editing messages.

### Ideal solution or implementation

Perhaps the `rateLimitOffset` option in the client options could optionally be a function, similar to `rejectOnRatelimit`. This could take in the route, or extra data, whichever works best, and return the number the offset should be in this context.

### Alternative solutions or implementations

The alternative is to implement your own external rate limiting, but that can become messy or repetitive depending on the bot.

### Other context

_No response_

### PR DESCRIPTION
**Pull request by didinele**
**Please describe the changes this PR makes and why it should be merged:**
Closes #9762

### CODE REVIEW

**Review by vladfrangu — APPROVED** [review]
Up for debate if we should maybe make GetRateLimitOffsetFunction async compat or not, I'd say nah, but otherwise LGTM

File: packages/rest/src/lib/utils/types.ts [hunk 1]
```diff
@@ -191,6 +191,11 @@ export interface RateLimitData {
  */
 export type RateLimitQueueFilter = (rateLimitData: RateLimitData) => Awaitable<boolean>;
 
+/**
+ * A function that determines the ratelimit offset for a given request.
```
Comment (Jiralite) [inline packages/rest/src/lib/utils/types.ts:195]:
```suggestion
 * A function that determines the rate limit offset for a given request.
```

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 2]
```diff
@@ -209,9 +210,11 @@ export class SequentialHandler implements IHandler {
 			let delay: Promise<void>;
 
 			if (isGlobal) {
+				const offset = normalizeRateLimitOffset(this.manager.options.offset, routeId.bucketRoute);
+
 				// Set RateLimitData based on the global limit
 				limit = this.manager.options.globalRequestsPerSecond;
-				timeout = this.manager.globalReset + this.manager.options.offset - Date.now();
+				timeout = offset - Date.now();
```
Comment (ckohen) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:217]:
```suggestion
				timeout = this.manager.globalReset + offset - Date.now();
```

You removed a bit too much. 👋 Global ratelimit

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 3]
```diff
@@ -341,13 +346,15 @@ export class SequentialHandler implements IHandler {
 			let timeout: number;
 
 			if (isGlobal) {
+				const offset = normalizeRateLimitOffset(this.manager.options.offset, routeId.bucketRoute);
+
 				// Set RateLimitData based on the global limit
 				limit = this.manager.options.globalRequestsPerSecond;
-				timeout = this.manager.globalReset + this.manager.options.offset - Date.now();
+				timeout = offset - Date.now();
```
Comment (ckohen) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:353]:
```suggestion
				timeout = this.manager.globalReset + offset - Date.now();
```

Ditto

**Review by ckohen — CHANGES_REQUESTED** [review]

**Review by ckohen — APPROVED** [review]

**Review by Jiralite — APPROVED** [review]

**Pull request merged**